core/metrics_manager.js

```diff
@@ -0,0 +1,385 @@
+ * (workspace comments and blocks).
+ *
+ * If the workspace does not have a fixed width and height then it is the
+ * metrics of the area that content can be placed. This area is computed by
+ * getting the rectangle around the top bounded elements on the workspace and
+ * adding padding to all sides.
+ * @param {boolean=} opt_getWorkspaceCoordinates True to get the content metrics
+ *     in workspace coordinates, false to get them in pixel coordinates.
+ * @return {!Blockly.MetricsManager.ContainerRegion} The
+ *     metrics for the content container.
+ * @public
+ */
+Blockly.MetricsManager.prototype.getContentMetrics = function(
+    opt_getWorkspaceCoordinates) {
+  var scale = opt_getWorkspaceCoordinates ? this.workspace_.scale : 1;
+  var viewSize = this.getViewMetrics(false);
+  var contentDimensions = this.getContentDimensions_(viewSize);
+
+  return {
+    height: contentDimensions.height / scale,
+    width: contentDimensions.width / scale,
+    top: contentDimensions.top / scale,
+    left: contentDimensions.left / scale,
+  };
+};
+
+/**
+ * Returns an object with all the metrics required to size scrollbars for a
+ * top level workspace.  The following properties are computed:
+ * Coordinate system: pixel coordinates, -left, -up, +right, +down
+ * .viewHeight: Height of the visible portion of the workspace.
+ * .viewWidth: Width of the visible portion of the workspace.
+ * .contentHeight: Height of the content.
+ * .contentWidth: Width of the content.
+ * .svgHeight: Height of the Blockly div (the view + the toolbox,
+ *    simple or otherwise),
+ * .svgWidth: Width of the Blockly div (the view + the toolbox,
+ *    simple or otherwise),
+ * .viewTop: Top-edge of the visible portion of the workspace, relative to
+ *     the workspace origin.
+ * .viewLeft: Left-edge of the visible portion of the workspace, relative to
+ *     the workspace origin.
+ * .contentTop: Top-edge of the content, relative to the workspace origin.
+ * .contentLeft: Left-edge of the content relative to the workspace origin.
+ * .absoluteTop: Top-edge of the visible portion of the workspace, relative
+ *     to the blocklyDiv.
+ * .absoluteLeft: Left-edge of the visible portion of the workspace, relative
+ *     to the blocklyDiv.
+ * .toolboxWidth: Width of the toolbox, if it exists.  Otherwise zero.
+ * .toolboxHeight: Height of the toolbox, if it exists.  Otherwise zero.
+ * .flyoutWidth: Width of the flyout if it is always open.  Otherwise zero.
+ * .flyoutHeight: Height of the flyout if it is always open.  Otherwise zero.
+ * .toolboxPosition: Top, bottom, left or right. Use TOOLBOX_AT constants to
+ *     compare.
+ * @return {!Blockly.utils.Metrics} Contains size and position metrics of a top
+ *     level workspace.
+ * @public
+ */
+Blockly.MetricsManager.prototype.getMetrics = function() {
+  var contentMetrics = this.getContentMetrics();
+  var viewMetrics = this.getViewMetrics();
```

Comment:
One concern I have is that we compute `viewMetrics` twice in `getMetrics` since there's a call for `getViewMetrics` inside of `getContentMetrics`. 
Maybe we can add an optional parameter to contentMetrics so that we can pass viewmetrics if we already have that computed.

Reply:
Ok done.